cdc/generation_service.hh

```diff
@@ -0,0 +1,136 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <seastar/core/semaphore.hh>
+
+#include "gms/i_endpoint_state_change_subscriber.hh"
+
+#include "metadata.hh"
+#include "generation.hh"
+
+namespace gms {
+    class feature_service;
+}
+
+namespace cdc {
+
+/* Managing CDC generations:
+ * - Creating a new generation when joining a cluster or performing a rolling upgrade.
+ * - Listening to generation changes.
+ * - Communicating generations between nodes.
+ *
+ * This service really runs its logic only on shard 0, but it stores and updates
+ * a cdc::metadata instance on each shard for efficient access by the log service.
+ */
+class generation_service : public enable_shared_from_this<cdc::generation_service>,
+                           public peering_sharded_service<cdc::generation_service>,
+                           public gms::i_endpoint_state_change_subscriber {
+
+    /* The highest generation timestamp that we know of.
+     * We keep this variable synchronized with our local tables (system.cdc_local)
+     * and the timestamp that we're gossiping (CDC_STREAMS_TIMESTAMP). */
+    std::optional<db_clock::time_point> _current_ts;
+
+    /* We update _current_ts together with gossiper and system keyspace.
+     * We want these three updates to happen atomically if multiple timestamp changes
+     * are detected concurrently (should not happen). */
+    semaphore _current_ts_sem{1};
+
+    cdc::metadata _metadata;
+
+    bool _stopped = false;
+
+    /* We will be updating the gossiper with the right CDC generation timestamp
+     * and listening for other nodes changing their gossiped timestamps. */
+    gms::gossiper& _gossiper;
+
+    /* We will be inserting and retrieving generation descriptions
+     * to/from the system_distributed.cdc_topology_description table,
+     * and updating the user-facing system_distributed.cdc_description table with appropriate
+     * lists of streams. */
+    sharded<db::system_distributed_keyspace>& _sys_dist_ks;
+
+    /* Some of our tasks will be sleeping. We need the abort source in case of shutting down. */
+    abort_source& _abort_source;
+
+    const db::config& _db_cfg;
+    const locator::token_metadata& _ring;
+
+    /* To keep the service from stopping if there are any asynchronous tasks still running,
+     * such as learning a generation or updating the streams description table. */
+    gate _async_gate;
+
+public:
+    // See comment in main.cc for the reason of taking sharded<db::sys_dist_ks>& instead of simply db::sys_dist_ks&.
+    generation_service(gms::gossiper&, sharded<db::system_distributed_keyspace>&,
+            abort_source&, const db::config&, const locator::token_metadata&);
+
+    future<> stop();
+    ~generation_service();
```

Comment:
Shouldn't this be virtual?

Reply:
Only gossiper holds a pointer to a base class of `generation_service` (it holds a pointer to the subscriber interface). But we `unregister_` this pointer in `generation_service::stop()`. So the pointer which destroys us is of type `generation_service*`.

Reply:
Adding virtual does not hurt.

Reply:
@elcallio this PR is no longer relevant, it is superseded by https://github.com/scylladb/scylla/pull/8101